Add initial free-threading page for the guide (#4577)
PyO3/pyo3: guide/src/free-threading.md

# Supporting Free-Threaded CPython

CPython 3.13 introduces an experimental build of CPython that does not rely on

> the global interpreter lock for thread safety. As of version 0.23, PyO3 also has

Reviewer: ```suggestion
the global interpreter lock (GIL) for thread safety. As of version 0.23, PyO3 also has
```
Maybe we want to introduce the acronym for the following section?

preliminary support for building rust extensions for the free-threaded Python
build and support for calling into free-threaded Python from Rust.

The main benefit for supporting free-threaded Python is that it is no longer
necessary to rely on rust parallelism to achieve concurrent speedups using
PyO3. Instead, you can parallelise in Python using the
[`threading`](https://docs.python.org/3/library/threading.html) module, and

> still expect to see see multicore speedups by exploiting threaded concurrency in

Reviewer 2: There is a `see see` typo.

Python, without any need to release the GIL. If you have ever needed to use
`multiprocessing` to achieve a speedup for some algorithm written in Python,
free-threading will likely allow the use of Python threads instead for the same

> workflow.

Reviewer: Maybe we want to also talk a bit about what PyO3/Rust brings to the table here? Maybe about the "fearless concurrency" point, having `Send`/`Sync` allowing to safely build APIs that can be used in this new context?

free-threaded Python. While many simple PyO3 uses, like defining an immutable
python class, will likely work "out of the box", there are currently some
limitations. 

## Many symbols exposed by PyO3 have `GIL` in the name

We are aware that there are some naming issues in the PyO3 API that make it
awkward to work in an environment where there is no GIL. We plan to change the
names of these types to deemphasize the role of the GIL in future versions of
PyO3, but for now you should remember that the use of the term `GIL` in
functions and types like `with_gil` and `GILOnceCell` is historical.

Instead, you can think about whether or not you a rust scope has access to a
Python **thread state** in `ATTACHED` status. See [PEP
703](https://peps.python.org/pep-0703/#thread-states) for more background about
Python thread states and status. In order to use the CPython C API in both the
GIL-enabled and free-threaded builds of CPython, you must own an attached
Python thread state. The `with_gil` function sets this up and releases the
thread state after the closure passed to `with_gil` finishes. Similarly, in both
the GIL-enabled and free-threaded build, you must use `allow_threads` in

> order to use rust threads. Both of `with_gil` and `allow_threads` tell CPython

Reviewer: I'm not quite sure how to interpret the "Rust threads" part. I don't think there is a problem in just spawning Rust thread an doing something different non Python related in the background. It's more about detaching the _current_ thread from it's interaction with the Interpreter and handing back control, if I got that correctly.

to put the Python thread state into `DETACHED` status. In the GIL-enabled build,
this is equivalent to releasing the GIL. In the free-threaded build, this unblocks

> CPython from triggering a stop-the-world for a garbage collection pass.

Reviewer: This is very valuable information, but also pretty technical. Maybe we should put this in a `<details>` tab and summarize that on a higher level, so that it's a bit easier to grasp for new users. Maybe that "GIL" refers to the interaction with the Python interpreter or something like that.

# Supporting Free-Threaded CPython

CPython 3.13 introduces an experimental "free-threaded" build of CPython that
does not rely on the [global interpreter
lock](https://docs.python.org/3/glossary.html#term-global-interpreter-lock)
(often referred to as the GIL) for thread safety. As of version 0.23, PyO3 also

> has preliminary support for building rust extensions for the free-threaded

Reviewer 3: Should we be capitalizing Rust? 🤔 (A few cases below also, if you agree...)

```suggestion
has preliminary support for building Rust extensions for the free-threaded
```

limitations.

## Many symbols exposed by PyO3 have `GIL` in the name

We are aware that there are some naming issues in the PyO3 API that make it
awkward to think about a runtime environment where there is no GIL. We plan to
change the names of these types to de-emphasize the role of the GIL in future
versions of PyO3, but for now you should remember that the use of the term `GIL`
in functions and types like `with_gil` and `GILOnceCell` is historical.

Instead, you can think about whether or not a rust thread is attached to a
Python interpreter runtime. See [PEP
703](https://peps.python.org/pep-0703/#thread-states) for more background about
how threads can be attached and detached from the interpreter runtime, in a
manner analagous to releasing and acquiring the GIL in the GIL-enabled build.

Calling into the CPython C API is only legal when an OS thread is explicitly
attached to the interpreter runtime. In the GIL-enabled build, this happens when
the GIL is acquired. In the free-threaded build there is no GIL, but the same C
macros that release or acquire the GIL in the GIL-enabled build instead ask the
interpreter to attach the thread to the Python runtime, and there can be many

> threads simultaneously attached.

Reviewer 3: I think we should add a sentence here along the lines of

"Similarly, in the GIL-enabled build PyO3 uses the `Python<'py>` type and the `'py` lifetime to signify attachment the global interpreter lock; in the freethreaded build these simply mean the thread is currently attached to the Python interpreter."

Author: I incorporated this and edited the surrounding paragraphs a bit. I think it reads better now, thanks for the suggestion.

The main reason for attaching to the Python runtime is to interact with Python
objects or call into the CPython C API. To interact with the Python runtime, the

> thread must register itself by attaching to the interpreter runtime.

Reviewer 3: Maybe mention `with_gil` here as the way to attach and the plan is that it will continue to exist under a new name?

leveraging the GIL to lock concurrent access from other threads. In
free-threaded Python there is no GIL, so you will need to replace this type with
some other form of locking. In many cases, a type from `std::sync::Atomic` or
a `std::sync::Mutex` will be sufficient.

Before:

```rust
# fn main() {
# #[cfg(not(Py_GIL_DISABLED))] {
# use pyo3::prelude::*;
use pyo3::sync::GILProtected;
use pyo3::types::{PyDict, PyNone};
use std::cell::RefCell;

static OBJECTS: GILProtected<RefCell<Vec<Py<PyDict>>>> =
    GILProtected::new(RefCell::new(Vec::new()));

Python::with_gil(|py| {
    // stand-in for something that executes arbitrary Python code
    let d = PyDict::new(py);
    d.set_item(PyNone::get(py), PyNone::get(py)).unwrap();
    OBJECTS.get(py).borrow_mut().push(d.unbind());
});
# }}
```

After:

```rust
# use pyo3::prelude::*;
# fn main() {
use pyo3::types::{PyDict, PyNone};
use std::sync::Mutex;

static OBJECTS: Mutex<Vec<Py<PyDict>>> = Mutex::new(Vec::new());

Python::with_gil(|py| {
    // stand-in for something that executes arbitrary Python code
    let d = PyDict::new(py);
    d.set_item(PyNone::get(py), PyNone::get(py)).unwrap();

>     // we're not executing Python code while holding the lock, so GILProtected
>     // was never needed

Reviewer 3: ```suggestion
    // as with any `Mutex` usage, lock the mutex for as little time as possible
    // in this case, we do it just while pushing into the `Vec`
```

    OBJECTS.lock().unwrap().push(d.unbind());
});
# }
```

If you are executing arbitrary Python code while holding the lock, then you will
need to use conditional compilation to use `GILProtected` on GIL-enabled Python
builds and mutexes otherwise. If your use of `GILProtected` does not guard the
execution of arbitrary Python code or use of the CPython C API, then conditional
compilation is likely unnecessary since `GILProtected` was not needed in the

> first place and instead rust mutexes or atomics should be preferred. Python 3.13
> introduces `PyMutex`, which releases the GIL while the lock is held, so that is
> another option if you only need to support newer Python versions.

Reviewer 3: Is this interaction between `PyMutex` and the GIL true? I thought it just released the GIL while blocking.

```suggestion
first place and instead rust mutexes or atomics should be preferred. Python 3.13
introduces `PyMutex`, which releases the GIL while the waiting for the lock, so that is
another option if you only need to support newer Python versions.
```

One option is that we add `MutexExt` traits to `pyo3::sync` which work for all Python versions, similar to #4513 

e.g. `.lock_py_attached(py)`

🤔

Author: Thanks for spotting the error, I applied your correction.

The extension trait is an interesting idea, it's certainly possible for PyO3 to detach from the GIL if a mutex is blocked. Definitely something to consider...

limitations.

## Many symbols exposed by PyO3 have `GIL` in the name

We are aware that there are some naming issues in the PyO3 API that make it
awkward to think about a runtime environment where there is no GIL. We plan to
change the names of these types to de-emphasize the role of the GIL in future
versions of PyO3, but for now you should remember that the use of the term `GIL`
in functions and types like `with_gil` and `GILOnceCell` is historical.

Instead, you can think about whether or not a rust thread is attached to a
Python **thread state**. See [PEP
703](https://peps.python.org/pep-0703/#thread-states) for more background about
Python thread states and status.

In order to use the CPython C API in both the GIL-enabled and free-threaded
builds of CPython, the thread calling into the C API must own an attached Python
thread state. In the GIL-enabled build the thread that holds the GIL by
definition is attached to a valid Python thread state, and therefore only one
thread at a time can call into the C API.

What a thread releases the GIL, the Python thread state owned by that thread is
detached from the interpreter runtime, and it is not valid to call into the

> CPython C API.

Author: Thanks for the suggestion, I've re-written this to talk about being attached to an "interpreter runtime" which is also vague but maybe a bit more suggestive than "thread state"?

> CPython C API.

Reviewer: For me this reads really nicely now ❤️ I think it paints a good, high level picture now with all the necessary information without being overwhelmingly technical 👍  Well done!

In the free-threaded build, more than one thread can simultaneously call into
the C API, but any thread that does so must still have a reference to a valid
attached thread state. The CPython runtime also assumes it is responsible for
creating and destroying threads, so it is necessary to detach from the runtime
before creating any native threads outside of the CPython runtime. In the
GIL-enabled build, this corresponds to dropping the GIL with an `allow_threads`

> call.

Author: You're right, it's weaker than that, see https://docs.python.org/3/c-api/init.html#non-python-created-threads.

I've updated this so now the restriction I'm referring to only refers to OS threads that need to interact with Python objects, the rayon example you linked to is messing with rust strings, so no need to be attached to a Python interpreter runtime.

> call.

Reviewer: Thanks for the link. That makes a lot of sense. In this case I think we're fine, because we don't allow the `Python` token to cross thread boundaries (`!Send`) and `Python::with_gil` (or `Python::attach` in the future?) will register the current thread correctly, right?

> call.

Author: Yup, `with_gil` is equivalent to calling `PyGILState_Ensure` and `PyGILStste_Release` before and after some C code that interacts with the C API.

(often referred to as the GIL) for thread safety. As of version 0.23, PyO3 also
has preliminary support for building Rust extensions for the free-threaded
Python build and support for calling into free-threaded Python from Rust.

If you want more background on free-threaded Python in general, see the [what's
new](https://docs.python.org/3.13/whatsnew/3.13.html#whatsnew313-free-threaded-cpython)
entry in the CPython docs, the [HOWTO
guide](https://docs.python.org/3.13/howto/free-threading-extensions.html#freethreading-extensions-howto)
for porting C extensions, and [PEP 703](https://peps.python.org/pep-0703/),
which provides the technical background for the free-threading implementation in
CPython.

The main benefit for supporting free-threaded Python is that both pure Python
code and code interacting with Python via the CPython C API can run
simultaneously on multiple OS threads.  This means it's much more
straightforward to achieve multithreaded parallelism in the Python layer, using
the `threading` module, without any need to ensure the GIL is released during
compute-heavy tasks, or any worries about [Amdahl's
law](https://en.wikipedia.org/wiki/Amdahl%27s_law) limiting parallel

> performance. If you have ever needed to use `multiprocessing` to achieve a

Reviewer 4: I don't understand the amdahl's law reference here, it's absolutely still relevant :-)

I think what you're trying to say is "in a GIL world, the value of releasing the GIL is bound by the portion of your code that drops the GIL (under Amdahl's law), while in the no GIL world, you can gain teh benefits to the full extent your code is parallel", but that's not clear from the text.

Author: I rewrote this to try to make the point about Amdahl's law more explicit. Thanks for pointing out that this was unclear.

Reviewer 4: Looks good

long-running tasks is no longer necessary to ensure other threads run, but you
should still detach from the interpreter runtime using [`Python::allow_threads`]
when doing long-running tasks that do not require the CPython runtime. The
garbage collector can only run if all threads are detached from the runtime (in
a stop-the-world state), so detaching from the runtime allows freeing unused
memory.

## Runtime panics for multithreaded access of mutable `pyclass` instances

If you wrote code that makes strong assumptions about the GIL protecting shared
mutable state, it may not currently be straightforward to support free-threaded
Python without the risk of runtime mutable borrow panics. PyO3 does not lock
access to Python state, so if more than one thread tries to access a Python
object that has already been mutably borrowed, only runtime checking enforces
safety around mutably aliased Rust variables the Python interpreter can
access. We believe that it would require adding an `unsafe impl` for `Send` or
`Sync` to trigger this behavior in code using PyO3. Please report any issues
related to runtime borrow checker errors on mutable pyclass implementations that

> do not make strong assumptions about the GIL.

Reviewer 4: Sorry, why do we think that `unsafe impl` is required?

Author: I'm not actually sure, that came from @davidhewitt

Reviewer 3: I think the point is that PyO3 uses `Send` and `Sync` bounds to rely on thread safety, but this is not really different from any other Rust code and also is relevant for everything, not just `#[pyclass]`.

Maybe we can replace this whole structure with something simpler like:

"`#[pyclass]` data is protected from concurrent access by a `RefCell`-like pattern of runtime borrow checking. Like a `RefCell`, PyO3 will raise exceptions (or in some cases panic) to enforce exclusive access for mutable borrows. With the GIL enabled it was unusual for these errors to be encountered in well structured code (using `allow_threads` or calling a `&self` `#[pymethod]` from a `&mut self` one could produce these errors)."

Reviewer 4: Yes, that's much clearer. Perhaps we could even include an example of a Python snippet likely to fail? `[threading.Thread(target=some_obj.some_method_that_takes_self_mut).start() for _ in range(16)]` (but cleaned up)

Author: I added a code example.

One question: do the code examples in the docs get run as doctests? I modeled what I added based on another example in the guide that has some python code, but looking at the nox configuration I don't see anything that checks the code examples in the guide.

Reviewer 3: That's a great question - you should add this new `.md` file to the list of `doctests!` in the bottom of `lib.rs` (around line 500).

It was always possible to generate panics like this in PyO3 in code that
releases the GIL with `allow_threads` (see [the docs on interior
mutability](./class.md#bound-and-interior-mutability),) but now in free-threaded
Python there are more opportunities to trigger these panics because there is no
GIL.

We plan to allow user-selectable semantics for mutable pyclass definitions in
PyO3 0.24, allowing some form of opt-in locking to emulate the GIL if that is
needed.

## `GILProtected` is not exposed

`GILProtected` is a PyO3 type that allows mutable access to static data by
leveraging the GIL to lock concurrent access from other threads. In
free-threaded Python there is no GIL, so you will need to replace this type with

> some other form of locking. In many cases, a type from `std::sync::Atomic` or

Reviewer 4: ```suggestion
some other form of locking. In many cases, a type from `std::sync::atomic` or
```
